### PR TITLE
Minor fixes for the CI

### PR DESCRIPTION
**Pull request by Geod24**

### CODE REVIEW

**Comment by TurkeyMan** [comment]
Thanks for hacking on this.
Just to be clear, only Windows passes unit tests, and that may possibly be only on my machine, with my particular compilers and environment ;)
It will also only work with the latest DMD, since I use copy constructors.

**Comment by Geod24** [comment]
> Just to be clear, only Windows passes unit tests

Only Windows compiles, actually :)
I do have a strong interest in this, and our targets are OSX / Linux, so I figured I would fix the CI for those targets. Now it compiles "as it should", as in, you have the right version of the compilers and it compiles the right code (before the "dependency" made `dub` fetch the package again, so it was not testing the code in PRs).

Will see if I can help with the Druntime binding for OSX. Also do I remember right that you had POSIX code in your previous PRs ? In any case, if you have to fight Travis in a future PR, you can always try to tag me and I'll take a look.

Current failure is:
```
../source/core/experimental/stdcpp/vector.d(54,5): Error: undefined identifier as_array
../source/core/experimental/stdcpp/vector.d(60,20): Error: undefined identifier size, did you mean alias size_t?
source/vector_test.d(63,23): Error: template instance `core.experimental.stdcpp.vector.std.vector!(int, allocator!int)` error instantiating
```

And the missing Windows SDK on AppVeyor.

So I think we could merge this ?

**Comment by TurkeyMan** [comment]
You're welcome to contribute. OSX uses clang, which is the next one I would do.

**Comment by TurkeyMan** [comment]
What do you mean POSIX code? This is strictly STL code. There's no platform code here.

**Comment by Geod24** [comment]
> What do you mean POSIX code? This is strictly STL code. There's no platform code here.

I mean that `as_array` and `size` are hidden behind a `version (CppRuntime_Microsoft)`, that's why the compilation fails on Travis.

**Comment by Geod24** [comment]
I've thrown in a binding for `type_info`.
By the way, are you sure of the `exception` derivative binding ?

**Comment by TurkeyMan** [comment]
I actually didn't write the `exception` code, and I've never tested to see if it works right... I assumed the original author might have done that ;) (...I suspect it doesn't work, it desperately needs unit tests)

**Comment by TurkeyMan** [comment]
`as_array` is in the bucket of functions that need a distinct implementation for each STL. Try and keep the code as close to the STL reference code as possible, so that it's possible to follow when comparing the codebases. i have a long-term plan that when all STL's are supported and working, we'll conflate all common elements into non-versioned functions and do a major simplification pass... but that's extremely hard to do up-front, because it makes it impossible to follow code when comparing against the respective C++ source code.

File: source/core/experimental/stdcpp/typeinfo.d [hunk 1]
```diff
@@ -143,5 +142,145 @@ else version (CppRuntime_Gcc)
...
+else version (CppRuntime_Clang)
+{
+    extern (C++, "std"):
+
+    static assert(__traits(classInstanceSize, type_info) == 16);
+
+    class type_info
+    {
+        void dtor0() {}
+        void dtor1() {}
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d]: This is old & dead style. Use `~this()` in 2019.

File: source/core/experimental/stdcpp/typeinfo.d [hunk 2]
```diff
@@ -143,5 +142,145 @@ else version (CppRuntime_Gcc)
...
+        void dtor1() nothrow {}
+        override const(char)* what() const nothrow;
+    }
+
+    class bad_typeid : exception
+    {
+    public:
+        this() nothrow;
+        void dtor0() nothrow {}
+        void dtor1() nothrow {}
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d]: ditto x2

**Comment by Geod24** [comment]
Updated:
```diff
diff --git a/source/core/experimental/stdcpp/typeinfo.d b/source/core/experimental/stdcpp/typeinfo.d
index af0a6b8..d368945 100644
--- a/source/core/experimental/stdcpp/typeinfo.d
+++ b/source/core/experimental/stdcpp/typeinfo.d
@@ -146,12 +146,13 @@ else version (CppRuntime_Clang)
 {
     extern (C++, "std"):

-    static assert(__traits(classInstanceSize, type_info) == 16);
+    static assert(__traits(classInstanceSize, type_info)  == 16);
+    static assert(__traits(classInstanceSize, bad_cast)   ==  8);
+    static assert(__traits(classInstanceSize, bad_typeid) ==  8);

     class type_info
     {
-        void dtor0() {}
-        void dtor1() {}
+        ~this() {}

     final:
         this(const ref type_info);
@@ -268,8 +269,7 @@ else version (CppRuntime_Clang)
     {
     public:
         this() nothrow;
-        void dtor0() nothrow {}
-        void dtor1() nothrow {}
+        ~this() nothrow {}
         override const(char)* what() const nothrow;
     }

@@ -277,8 +277,7 @@ else version (CppRuntime_Clang)
     {
     public:
         this() nothrow;
-        void dtor0() nothrow {}
-        void dtor1() nothrow {}
+        ~this() nothrow {}
         override const(char)* what() const nothrow;
     }
 }
```

File: source/core/experimental/stdcpp/typeinfo.d [hunk 3]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+
+    final:
+        this(const ref type_info);
+        // identity assignment operator overload is illegal
+        ref type_info opAssign_(const ref type_info);
+
+        version(LIBCPP_HAS_NONUNIQUE_TYPEINFO)
+        {
+            pragma(inline, true) extern(D)
+            int __compare_nonunique_names(const ref type_info __arg) const nothrow
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:165]: What's this? I can't see it in the libc++ source. Link?

File: source/core/experimental/stdcpp/typeinfo.d [hunk 4]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+        ~this() {}
+
+    final:
+        this(const ref type_info);
+        // identity assignment operator overload is illegal
+        ref type_info opAssign_(const ref type_info);
+
+        version(LIBCPP_HAS_NONUNIQUE_TYPEINFO)
+        {
+            pragma(inline, true) extern(D)
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:164]: I think you can stick `extern(D):` at the top of the class.

File: source/core/experimental/stdcpp/typeinfo.d [hunk 5]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+        }
+
+        version (Windows)
+        {
+            struct {
+                char*   __undecorated_name;
+                char[1] __decorated_name;
+            }
+
+            int __compare(const ref type_info __rhs) const nothrow;
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:178]: Where's the body?
```c++
int std::type_info::__compare(const type_info &__rhs) const _NOEXCEPT {
  if (&__data == &__rhs.__data)
    return 0;
  return strcmp(&__data.__decorated_name[1], &__rhs.__data.__decorated_name[1]);
}
```

File: source/core/experimental/stdcpp/typeinfo.d [hunk 6]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+        {
+            const(char)* name() const nothrow;
+
+            pragma(inline, true) extern(D)
+            bool before(const ref type_info __arg) const nothrow
+            {
+                return __compare(__arg) < 0;
+            }
+
+            size_t hash_code() const nothrow;
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:202]: No body?

Comment (TurkeyMan) [reply source/core/experimental/stdcpp/typeinfo.d:202]: Ohhh, they're not template functions! Yeah okay, extern is fine.

File: source/core/experimental/stdcpp/typeinfo.d (hunk 5 above)
Comment (TurkeyMan) [reply source/core/experimental/stdcpp/typeinfo.d:178]: Sorry, I didn't notice they were not template functions.

File: source/core/experimental/stdcpp/typeinfo.d (hunk 3 above)
Comment (Geod24) [reply source/core/experimental/stdcpp/typeinfo.d:165]: https://opensource.apple.com/source/clang/clang-800.0.42.1/src/projects/libcxx/include/typeinfo.auto.html

**Comment by TurkeyMan** [comment]
It'd be nice if this has a unittest. I usually pass one from D by-value to C++, and then call the methods and make sure they work, then pass one from C++ by-value to D, then call the methods and make sure they work.
Passing in each direction confirms the ABI.

File: source/core/experimental/stdcpp/typeinfo.d (hunk 3 above)
Comment (TurkeyMan) [reply source/core/experimental/stdcpp/typeinfo.d:165]: Hmm, I'd rather refer to mainline LLVM: https://github.com/llvm-mirror/libcxx/blob/master/include/typeinfo
Have Apple forked it? Does apple have patches?

File: source/core/experimental/stdcpp/typeinfo.d [hunk 7]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+    static assert(__traits(classInstanceSize, bad_typeid) ==  8);
+
+    class type_info
+    {
+        ~this() {}
+
+    final:
+        this(const ref type_info);
+        // identity assignment operator overload is illegal
+        ref type_info opAssign_(const ref type_info);
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:160]: Is there any reason to declare this at all? It's not virtual...

File: source/core/experimental/stdcpp/typeinfo.d [hunk 8]
```diff
@@ -143,5 +142,144 @@ else version (CppRuntime_Gcc)
...
+                if (!((__type_name & __arg.__type_name) & _LIBCPP_NONUNIQUE_RTTI_BIT))
+                    return false;
+                return __compare_nonunique_names(__arg) == 0;
+            }
+        }
+        else
+        {
+            pragma(inline, true) extern(D)
+            const(char*) name() const nothrow
+            { return __type_name; }
```
Comment (TurkeyMan) [inline source/core/experimental/stdcpp/typeinfo.d:252]: Please don't do this weird apple brace thing. I'd like all the code to match.

File: source/core/experimental/stdcpp/typeinfo.d (hunk 3 above)
Comment (Geod24) [reply source/core/experimental/stdcpp/typeinfo.d:165]: To be fair, I just looked at what's installed on my machine, not the actual llvm-mirror code

File: source/core/experimental/stdcpp/typeinfo.d (hunk 7 above)
Comment (Geod24) [reply source/core/experimental/stdcpp/typeinfo.d:160]: Right that's a leftover (and it wouldn't link). Will remove.

**Comment by Geod24** [comment]
So, provided your recent advance in Druntime, do you have plans for this repository ?
Sorry it kinda fell off the radar...

**Comment by TurkeyMan** [comment]
I've been juggling both repos, but I had to work on CI either way, so I put my time into the druntime CI, since it needs to get sorted regardless.

**Pull request closed without merging**